readme.md
<p align="center">
 <h2 align="center">Github Repo Gist</h2>
 <p align="center">This project imitates the Gist functionality provided by github but with a little twist, it works for
    repository source files.</p>

  <p align="center">
    <a href="https://github.com/thefallenmerc/git-repo-gist/issues">
      <img alt="Issues" src="https://img.shields.io/github/issues/thefallenmerc/git-repo-gist?color=0088ff" />
    </a>
    <a href="https://github.com/thefallenmerc/git-repo-gist/pulls">
      <img alt="GitHub pull requests" src="https://img.shields.io/github/issues-pr/thefallenmerc/git-repo-gist?color=0088ff" />
    </a>
  </p>

  <p align="center">
    <a href="https://git-repo-gist.herokuapp.com/framed/thefallenmerc/git-repo-gist/master/app.js">View Demo</a>
    ·
    <a href="https://github.com/thefallenmerc/git-repo-gist/issues">Report Bug</a>
    ·
    <a href="https://github.com/thefallenmerc/git-repo-gist/issues">Request Feature</a>
  </p>
</p>

# Index

- [Github Repo Gist](#github-stats-card)
- [How to use it](#how-to-use-it)
- [Customization](#customization)
- [Features](#features)

### Github Repo Gist

Github provides you with the functionality to make and embed a gist anywhere on the internet, but there's something missing, what if we want to share a gist of code from existing repo without the need of creating a new gist for it?

This is where this code comes in.

### How to use it

There are few simple steps to get gist out of code - 

- Get the `uri` to the `raw` file, for eg `https://raw.githubusercontent.com/thefallenmerc/git-repo-gist/master/app.js`
- extract the part starting from your username till end, (everything after `https://raw.githubusercontent.com/`), we get `thefallenmerc/git-repo-gist/master/app.js`
- Append this to our base url with any of the option available in between, the solution is hosted at [https://git-repo-gist.herokuapp.com/](https://git-repo-gist.herokuapp.com/), thus we get `https://git-repo-gist.herokuapp.com/<option>/thefallenmerc/git-repo-gist/master/app.js`, where option can have the following value -
    - __gist__: gives back and actual gist script
    - __framed__: gives back an iframe with the gist in it

<!-- >> Options: `&hide=["stars","prs","issues","contribs"]` -->

### Customization

You can now specify the portion of code you actually want to use in gist, instead of all of the file.

This can be done using URL query params to the final url, for eg -

> ?from=1&to=5

The `from` and `to` are the line numbers, and also included. 

the final URL will now be - `https://git-repo-gist.herokuapp.com/<option>/thefallenmerc/git-repo-gist/master/app.js?from=1&to=5`

### Features

- Two modes available
    - Script and Iframe
- Supported Languages (I suppose all the popular languages are supported but the list below is for sure)
    - typescript
    - javascript
    - dart
    - json
    - php
  

---
---


Contributions are welcomed! <3

Made with :heart: and javascript.

changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0] - 2020-07-18

### Added

- Change log
- Constrainted gist is now possible

### Changed

- Controller structure


helpers/main.helper.js
const axios = require("axios");
const Prism = require('prismjs');
const loadLanguages = require('prismjs/components/');

const serverBase = process.env.NODE_ENV === 'production' ? "https://git-repo-gist.herokuapp.com/" : "http://localhost:3000/";
const css = serverBase + "stylesheets/style.css";
const base = "https://raw.githubusercontent.com";

// list of janguages
const extList = {
    css: 'css',
    js: 'javascript',
    jsx: 'javascript',
    ts: 'typescript',
    tsx: 'typescript',
}


const helper = {
    /**
     * Contraints the number of line of the diplayed code to
     * 
     * @param code: code to be constrained
     * @param start: starting line, nullable 
     * @param end: ending line, nullable 
     */
    getConstrained: (code = "", start, end) => {
        // if start is invalid
        if (isNaN(start) || start < 1) {
            start = 1;
        }
        // if end is invalid or end is smaller than start
        if (isNaN(end) || end < start) {
            end = 10000; // default end to 10000, so that we dont have to worry about it
        }
        end = parseInt(end);
        // slice the code accordingly
        code = code.split("\n").slice(start - 1, end).join("\n");
        return code;
    },
    /**
     * Creates template for the highlighted code
     * @param highlight the highlighted code by prismjs
     * @param gist the url of actualy file on github
     */
    getTemplate: (highlight, gist) => {
        return `
        <div class="gist">
        <div class="gist-content">
        <pre class="line-numbers"><code>${highlight}</code></pre>
        </div>
        <div class="gist-meta">
        <a href="http://github.com${gist.replace('/master/', '/blob/master/')}">${gist}</a>
        provided with with &#10084; by <a href="${serverBase}">Github Repo Gist</a>
        <a href="${base + gist}" style="float: right;">raw</a>
        </div>
        </div>
        `;
    },
    /**
     * @param gist template whose gist will be made
     */
    makeGistScript: (gist) => {
        return `
        document.write('<link rel="stylesheet" href="${css}">')
        document.write(${JSON.stringify(gist)})
        `;
    },
    getRawFileContent: (fileUrl) => {
        // raw github url
        const remoteUrl = base + '/' + fileUrl;
        // get the raw content
        return axios.get(remoteUrl);
    },
    /**
     * Function returns rendered template for data that is provided
     * @param gist relative url of the raw file
     * @param data content of the file
     * @param start starting constraint for the file
     * @param end ending constraint for the file
     */
    getTemplateWithHighlightedCode: (gist, data, start, end) => {
        try {
            // extension
            const ext = gist.substr(gist.lastIndexOf('.') + 1);
            // get language or set extension
            const lang = extList[ext] || ext;
            loadLanguages([lang]);
            // check for offset and limit
            if (start || end) {
                code = helper.getConstrained(code, start, end);
            }
            // highlight content
            const highlight = Prism.highlight(code, Prism.languages[lang], lang);
            // create gist template
            const render = helper.getTemplate(highlight, gist);
            // return the created gist
            return helper.makeGistScript(render)
        } catch(error) {
            console.error(error);
            code = 'Something went wrong!';
            const render = helper.getTemplate(code, gist);
            return helper.makeGistScript(render);
        }
    }
}

module.exports = helper

controllers/frame.controller.js
const { Remarkable } = require('remarkable');
const Prism = require('prismjs');
const axios = require('axios');
const helper = require("../helpers/main.helper");

// list of janguages
const extList = {
    css: 'css',
    js: 'javascript',
    jsx: 'javascript',
    ts: 'typescript',
    tsx: 'typescript',
}

module.exports = function (req, res) {
    // gist url without github.com
    const gist = req.params[0];
    // extension
    const ext = gist.substr(gist.lastIndexOf('.') + 1);
    // get language or set extension
    const lang = extList[ext] || ext;
    // get raw file
    let finalizedGist = "";
    helper.getRawFileContent(gist)
        .then(response => {
            code = lang.toLowerCase() === 'json' ? JSON.stringify(response.data, null, 2) : response.data;
            // get finalized gist
            finalizedGist = helper.getTemplateWithHighlightedCode(gist, code, req.query.from, req.query.to);
        })
        .catch(error => {
            console.error(error);
            code = 'Something went wrong!';
            // get finalized gist
            finalizedGist = helper.getTemplateWithHighlightedCode(gist, code, req.query.from, req.query.to);
        }).finally(() => {
            // render it
            res.render('frame', { title: gist, gist: finalizedGist });
        });
}

controllers/gist.controller.js
const { Remarkable } = require('remarkable');
const md = new Remarkable();
const Prism = require('prismjs');
const axios = require('axios');
const loadLanguages = require('prismjs/components/');

const base = "https://raw.githubusercontent.com";
const serverBase = process.env.NODE_ENV === 'production' ? "https://git-repo-gist.herokuapp.com/" : "http://localhost:3000/";
const css = serverBase + "stylesheets/style.css";

// list of janguages
const extList = {
    css: 'css',
    js: 'javascript',
    jsx: 'javascript',
    ts: 'typescript',
    tsx: 'typescript',
}

// create gist template
function getTemplate(highlight, gist) {
    return `
    <div class="gist">
        <div class="gist-content">
            <pre class="line-numbers"><code>${highlight}</code></pre>
        </div>
        <div class="gist-meta">
            <a href="http://github.com${gist.replace('/master/', '/blob/master/')}">${gist}</a>
             provided with with &#10084; by <a href="${serverBase}">Github Repo Gist</a>
             <a href="${base + gist}" style="float: right;">raw</a>
        </div>
    </div>
`;
}

// make gist
function makeGistScript(gist) {
    return `
    document.write('<link rel="stylesheet" href="${css}">')
    document.write(${JSON.stringify(gist)})
    `;
}

module.exports = function (req, res) {
    // gist url without github.com
    const gist = req.params[0];
    // raw github url
    const remoteUrl = base + '/' + gist;
    // extension
    const ext = gist.substr(gist.lastIndexOf('.') + 1);
    // get language or set extension
    const lang = extList[ext] || ext;
    loadLanguages([lang]);
    // get the raw content
    axios.get(remoteUrl)
        .then(response => {
            try {
                code = lang.toLowerCase() === 'json' ? JSON.stringify(response.data, null, 2) : response.data;
                // highlight content
                const highlight = Prism.highlight(code, Prism.languages[lang], lang);
                // create gist template
                const render = getTemplate(highlight, gist);
                // send rendered
                res.end(makeGistScript(render));
            } catch (e) {

                code = 'Something went wrong!';
                // highlight content
                // const highlight = Prism.highlight(JSON.stringify(code, null, 2), Prism.languages.javascript, 'javascript');
                // create gist template
                const render = getTemplate(code, gist);
                // send rendered
                res.end(makeGistScript(render));
            }
        })
        .catch(error => {
            console.log(error);
            code = `
    // Github Repository To Gist Converter
    
    // this utility can be used to create github like gists using source files from any github repo;

    // head on to github and select any source file from any repository and click on the raw view

    // for eg https://raw.githubusercontent.com/thefallenmerc/notebook/master/package.json
    
    // remove the https://raw.githubusercontent.com/ and append it to https://gitrepogist.web.app/ which forms the
    
    // repo gist url - https://gitrepogist.web.app/thefallenmerc/notebook/master/package.json
    
    `;
            // highlight content
            const highlight = Prism.highlight(JSON.stringify(code, null, 2), Prism.languages[lang], lang);
            // create gist template
            const render = getTemplate(highlight, gist);
            // send rendered
            res.set('Content-Type', 'application/javascript');
            res.end(makeGistScript(render));
        })
}